Source/WebCore/rendering/style/StyleGradientImage.cpp

```diff
@@ -186,12 +215,10 @@ RefPtr<Image> StyleGradientImage::image(const RenderElement* renderer, const Flo
 
 bool StyleGradientImage::knownToBeOpaque(const RenderElement& renderer) const
 {
-    bool hasColorFilter = renderer.style().hasAppleColorFilter();
+    auto& style = renderer.style();
+    bool hasColorFilter = style.hasAppleColorFilter();
```

Comment:
I think the issue with the failing ::first-line test is that these renderer taking functions (most importantly image()) should be using `firstLineStyle()` instead of `style()` when appropriate. Whether we are on the first line could be passed as an additional boolean argument, or style itself could be passed (though at least some of these use the renderer for other things too so it would also be needed).  

The only place (?) where the first-line bit needs to be passed is in `BackgroundPainter::paintFillLayer()` where you would test with something like `bool isFirstLine = box && box->lineBox()->isFirst();`

Reply:
Thanks Antti! To keep things a bit sane, I am going to do that in a follow up, and make a small revert here to do the currentColor lookup in the old place. Filed https://bugs.webkit.org/show_bug.cgi?id=247127 for this.